docs/blueprints/slow-down-and-retry.md

```diff
@@ -0,0 +1,91 @@
+
+In case of `Too Many Requests` during download operation we should be able to retry single Metainfo API requests. If downloading segment from storage nodes will fail we should retry downloading this single segment. Like with uploads we shouldn't retry segment download on every error.
+
+As an addition to retry logic we need to implement at least basic backoff mechanism that will increase the waiting time (between retries) up to a certain threshold. This will prevent from overwhelming satellite with retry requests. For our needs exponential backoff looks to be reasonable solution. 
+
+As an alternative for implementing exponential backoff on client side we can also consider to return with `To Many Request` error information when client should try to repeat request. The HTTP protocol defines that is such situation server may include a `Retry-After` header ([429](https://httpstatuses.com/429)). We can use similar approach and return delay value more attached to current satellite load.
+
+To avoid satellite side larger development for providing accurate `Retry-After` values we may start start with client side exponential backoff implementation and later migrate into retry interval defined by satellite.
+
+Additional rules:
+* Every retried request should be processed by backoff logic.
+* Retry and slow mechanism should be transient from libuplink consumer perspective except several necessary configurable options (see Implementation) and logging (see next point).
+* Every retry should be logged (severity WARNING or lower) to give details why retry occurs and how long it will take to repeat request.
+* Applying retry and slow down logic shouldn't degrade libuplink overall performance and increase number of round trips to satellite.
+
+## Implementation
+
+Most of client side logic should be placed around `private/metainfo/client.go`. This is main and only client for making calls to Metainfo API. Ideally retry and slow down logic for this API should transparent for client consumers and no implementation details should be leaking.
+
+Implementation for upload process is placed in `private/storage/segments/store.go` and there we should adjust code base to support retrying single segment upload.
+
+Implementation can be divided into several steps:
+1. Implement retry and slow down for Metainfo API requests, except `Batch` request
+2. Implement retry and slow down for Metainfo API `Batch` requests
+3. Implement retry and slow down for failed segment upload
+4. Implement retry and slow down for failed segment download
+
+Client side options:
+* initial interval
+* max elapsed time or number of retries ??
+* multipliter
+* randomization factor
```

Comment:
it's usually called jitter

Reply:
I took names from http://googleapis.github.io/google-http-java-client/exponential-backoff.html but I'm ok with jitter too

Reply:
Ah, the randomization factor might have a different meaning here

Reply:
```
sleep(interval + interval * randomization-factor * random(-1, 1))
sleep(interval + random(0, jitter))
```

Reply:
its `randomized_interval =
 retry_interval * (random value in range [1 - randomization_factor, 1 + randomization_factor])`

Reply:
Isn't that the same as the first line?

Reply:
I posted my comment without knowing about yours :)